### ISSUE TITLE
Handle embedded Tweets

### ISSUE THREAD

**Issue by ghostwords** [EFForg/privacybadger#1393]
Our ShareMeNot-based social button replacement logic was built specifically for ... social buttons. [SoundCloud support was added recently](https://github.com/EFForg/privacybadger/pull/749/files?w=1) so we could probably hack in embedded Tweets in the same way, but should we? We should consider usability and privacy angles. Ideally our approach would eventually cover embedded YouTube videos and the like.

Privacy-wise, it would be best to avoid loading anything at all from Twitter until requested by the user. We get this from existing widget replacement, right?

Usability-wise, the user would (at minimum) like to see the fact there is an embedded Tweet and would like to be able to see all Tweets on the page by clicking one of these Tweets. We should get this from the current system.

This can be improved by letting people permanently allow embedded Tweets (for the site).

Another improvement would be to show the contents of the Tweet without violating the user's privacy. How feasible is this? Seems doable with text-only Tweets (assuming the text is included on the first-party page directly, not in a Twitter iframe), but unclear with multimedia (image/video) Tweets.

**Comment by terrorist96** [comment]
Ghostery blocks Twitter by default as well, but the text of the tweets are still visible. See example:
![image](https://cloud.githubusercontent.com/assets/12802056/26503694/c36eb2ea-420f-11e7-8d56-893ace33bbd4.png)
vs
![image](https://cloud.githubusercontent.com/assets/12802056/26503735/dcdbacba-420f-11e7-8e51-e26f360ca5fc.png)

**Comment by cowlicks** [comment]
Here is my proposed containment strategy. Twitter's embedded tweets look like this:
```html
<blockquote class="twitter-tweet" data-lang="en">... tweet info ... </blockquote>
<script async src="//platform.twitter.com/widgets.js" charset="utf-8"></script>
```

We remove this from the dom, and insert an iframe:
```html
<iframe sandbox="allow-scripts" src="tweetjail.html"></iframe>
```
The "`sandbox`" attribute blocks the iframe from accessing cookies ([more here](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/iframe)), but allow it to run scripts that are contained to the iframe.

`tweetjail.html` is simply:
```html
<head>
  <script src="tweetjail.js"></script>
</head>
<body></body>
```

`tweetjail.js` listens for the tweet info from the parent, then loads the tweet, then sends info back to the parent so the frame can be set to the proper size. It should look roughly like:
```javascript
function receive(e) {
  // take tweet info  and insert it into `tweetjail.html`
}
window.addEventListener("message", receive, false);
// observe tweetjail dom and send info to parent about height/width of the tweet
// postMessage( iframe.contentWindow.document.body.scrollWidth, parent);
```

Then the parent listens for the height/width and resize's the iframe accordingly.

The part that is currently missing is how to catch the embedded tweet before it loads `platform.twitter.com/widgets.js`. I'll look at `socialwidgets.js` or other extensions to see what they do. Thoughts?

**Comment by ghostwords** [comment]
Sandboxed iframes seem meant to prevent some types of security issues associated with embedding third-party content on websites. How do they help us protect user privacy?

**Comment by cowlicks** [comment]
@ghostwords code running in the iframe cannot access the browser's cookies, or any part of the context from where the iframe is. It won't know the origin where it is loaded, or be able to see the DOM. But fingerprinting might still be possible.

**Comment by ghostwords** [comment]
WIP pull request here: #1404

**Comment by ghostwords** [comment]
Implemented in 4cf05333257536ae0fb93ba7cd1799cc23b02b0c and #2926.
